File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head, n):

        dummy = ListNode(0)
        dummy.next = head

        left = dummy
        right = head

        # Move right n steps ahead
        for _ in range(n):
            right = right.next

        # Move both until right reaches the end
        while right:
            left = left.next
            right = right.next

        # Remove the nth node from the end
        left.next = left.next.next

        return dummy.next

File: Data Structures & Algorithms/sqrtx/submission-0.java
class Solution {
    public int mySqrt(int x) {
        if (x < 2) {
            return x;
        }

        long left = 1;
        long right = x;
        int result = 1;

        while (left <= right) {
            long mid = left + (right - left) / 2;

            if (mid * mid <= x) {
                result = (int) mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return result;
    }
}